add travisci example (Ubuntu 14.04 and Mac OS X 10.11)

enricosada opened the pull request:
ref https://github.com/dotnet/cli/issues/2113

it add an example travisci to build:
- osx 10.11
- ubuntu 14.04

It use .NET Core SDK binaries, so it's pretty much curl sdk | unzip to a subdirectory `.dotnetsdk` and add that to PATH

I used the .net core sdk binaries instead of pkg because are easier to test locally, no need for sudo etc.
The script from dotnet/cli repo is not used because it's easier to just download and unzip, and also it's easier to understand.

i am pretty sure for osx deps (because of docs about brew)

But i am not sure about ubuntu (see apt deps) because these were needed some months ago, maybe now are not.

It use `generic` as travis language, so it's possibile to copy before_install, install in any travis, not only `csharp`.

/cc @blackdwarf @akoeplinger

enricosada committed:
add travisci example (Ubuntu 14.04 and Mac OS X 10.11)

travisci now support .net core sdk with `dotnet` key on language `csharp`

blackdwarf commented:
@enricosada good one. Can I bother you with a few questions?

- Why not use apt-get for Ubuntu? Would that be "bad" in some way? I thought Travis adds users to the sudo group on each build image.
- What is hard about the script?

enricosada commented:
> Why not use apt-get for Ubuntu? Would that be "bad" in some way? I thought Travis adds users to the sudo group on each build image.

I got mixed experiences with packages and installer during development of dotnet/cli, i know preview1 is more stable now.
I really like sdk binaries atm because i do a lot of side by side installation. Lot of people is new of .net sdk and has issues about installing a machine wide package. Like that can try locally if something doesnt work. just that.
Also sometimes is needed to run exactly the same version of ci server of a component. If the component installation doesnt mess with my machine installation, i can easier try it.

And it's the same install for osx and ubuntu atm, because `brew install dotnet` is missing atm.


> What is hard about the script?

I used to execute the script before in travis/appveyor. i can do that like [there](https://github.com/enricosada/fsharp-dotnet-cli-samples/blob/dd2e6e4515de9622f4f21da46148ec1b2c0f7c8f/.travis.yml#L39) also with a oneliner if you want.

But there is little value to use it for me. it's just curl + unzip. Maybe in windows is better the script because i cannot use the better oneliner of powershell 5, but it's easy to follow.

If i use the script:

- i need to pin a version instead of repo branch because i got bite before when changed ( beta -> preview channel ), some issue bash, etc (i know are fixed, but still i fear that)
- the script help to set the dotnet in PATH, but with ci it's useless, because PATH changes inside a script are lost (scope), i need to anyway set the manually the PATH
- The really nice stuff about the script was when it created the url for `Latest` because directory inside package url was `Latest` but filename part was `latest` (lowercase). really annoying. It's nice to have a build matrix like [there](https://github.com/enricosada/fsharp-dotnet-cli-samples/blob/dd2e6e4515de9622f4f21da46148ec1b2c0f7c8f/.travis.yml#L24-L27), and was annoying to compose the url for curl | unzip.


Anyway, as you wish, np for me about one way or another, i can update both pr.

enricosada commented:
anyway @blackdwarf can you pls enable both travisci and appveyor build of this repo? so we can test these pr before merge.
both travis and appveyor has an option about 'dont build if config file is missing' so it's not an issue with other commit.

richlander commented:
Same questions as #461.

enricosada commented:
> This is just the build script. How will people know how to use this? I'm thinking that an associated help doc explaining this would be very useful.

People need that as example how to install .net core sdk in appveyor, not how to do build.
Everyone has different build script (fake, rake, cake, shell), but all they need is the .net core sdk installed and in PATH, actual command for build (restore, pack, etc) are up to developer. 
But if they already use travis, that's standard.

> Does this file need to be in the root of the repo? If we had a samples directory, could it be in there?

Yes by design of travis that's the default where it's expected (and where developer search it). 
Also, if you enable travis for this repo (it's free for open source), we can test this script every commit for free, and also (if you want), can build all the examples in this repo.

> Will you update this when it needs updating?

sure, i use .net core sdk for open source project, that's really easy to update it
Also when @akoeplinger update the travis `csharp` image, i'll update this script too, but it's nice to know it's possibile to install .net core sdk in a different image (for example i have another language image but i want to install .net core)

am11 commented on .travis.yml:
```diff
@@ -0,0 +1,45 @@
+before install:
+  - if [[ "$TRAVIS_OS_NAME" == "osx" ]]; then brew update               ; fi
+  - if [[ "$TRAVIS_OS_NAME" == "osx" ]]; then brew install openssl      ; fi
+  - if [[ "$TRAVIS_OS_NAME" == "osx" ]]; then brew link --force openssl ; fi
```
nit: can be wrapped in one `if` like this: https://github.com/maxmind/GeoIP2-dotnet/blob/bc0a878/.travis.yml#L31-L35

am11 replied:
C++ version in different repo: https://github.com/maxmind/MaxMind-DB-Reader-dotnet/blob/1701f6a/.travis.yml :)

blackdwarf commented:
@enricosada the same request as for the #461 . Can you add this to the "Using CLI with CI" doc?

enricosada commented:
i'll do that now @blackdwarf

enricosada commented:
done @blackdwarf @cartermp .

Pls @mairaw can you review this too? I really appreciated the review of the other pr

blackdwarf commented on docs/core/tools/using-ci-with-cli.md:
```diff
@@ -64,7 +64,44 @@ in the [CLI repo](https://github.com/dotnet/core/blob/master/Documentation/prere
 ### TravisCI
-**TODO**
+
+The [travis-ci](https://travis-ci.org/) doesn't have the .NET Core SDK preview1 installed by default, but can be added to any
```
This needs to become "Preview 2".

blackdwarf commented on docs/core/tools/using-ci-with-cli.md:
```diff
@@ -64,7 +64,44 @@ in the [CLI repo](https://github.com/dotnet/core/blob/master/Documentation/prere
+    - os: linux # Ubuntu 14.04
+      dist: trusty
+      sudo: required
+      env: DOTNET_SDK_URL=https://go.microsoft.com/fwlink/?LinkID=798405 # v1.0.0-preview1
```
Change for preview 2.

blackdwarf commented on docs/core/tools/using-ci-with-cli.md:
```diff
@@ -64,7 +64,44 @@ in the [CLI repo](https://github.com/dotnet/core/blob/master/Documentation/prere
+      env: DOTNET_SDK_URL=https://go.microsoft.com/fwlink/?LinkID=798405 # v1.0.0-preview1
+    - os: osx # OSX 10.11
+      osx_image: xcode7.2
+      env: DOTNET_SDK_URL=https://go.microsoft.com/fwlink/?LinkID=798404 # v1.0.0-preview1
```
Same as above.

blackdwarf commented on docs/core/tools/using-ci-with-cli.md:
```diff
@@ -64,7 +64,44 @@ in the [CLI repo](https://github.com/dotnet/core/blob/master/Documentation/prere
+The url ( configured in the `DOTNET_SDK_URL` environment variable ) can also be set to any binary
+package defined in the [dotnet/cli readme](https://github.com/dotnet/cli). 
+
+That's useful to install a specific version, for example `1.0.0-preview1-002702`
```
Again, change to preview 2.

blackdwarf commented:
@enricosada thanks much for doing this! :smile: I've left some comments, mostly around the links and the versions. As Preview 2 of the SDK is going live today, we should change to point to there.

enricosada commented:
yes, i'll update it when the `preview2` is released, useless merge the `preview1`

enricosada commented:
@blackdwarf if you want we can fix the links after the `preview2` released, less diff in docs.

cartermp commented:
@enricosada Once you get the links updated for Preview 2 we can merge this 😄.

enricosada commented:
@richlander @cartermp @blackdwarf  sry for delay, the previous install script was ok, i had some intermmittent issue with ssl/crypy issue after install from script. dunno why, but i was trying to cleanup more.

now travis support the `dotnet` key in the language `csharp`, so is really really better.

rebased to master, updated docs, and added example `.travis.yml`

qinezh commented:
Open Publishing Build Service: The pull request content has been published and here are some sample preview links:
- Docset 1: [core-docs](https://stage.docs.microsoft.com/en-us/dotnet/core/api/system.security.cryptography.cnguipolicy?branch=pr-en-us-460)

mairaw commented on docs/core/tools/using-ci-with-cli.md:
```diff
@@ -64,7 +64,17 @@ in the [CLI repo](https://github.com/dotnet/core/blob/master/Documentation/prere
 ### TravisCI
-**TODO**
+
+The [travis-ci](https://travis-ci.org/) can be configured to install the .NET Core SDK using the language `csharp` and `dotnet` key
```
nit: missing period on the end of the sentence

also wondering if this would read better like this:
...using the `csharp` language and the `dotnet` key.

mairaw commented on docs/core/tools/using-ci-with-cli.md:
````diff
@@ -64,7 +64,17 @@ in the [CLI repo](https://github.com/dotnet/core/blob/master/Documentation/prere
+```
+
+Travis can run both `osx` (OSX 10.11) and `linux` ( Ubuntu 14.04 ) job in a build matrix, see [example .travis.yml](https://github.com/dotnet/core-docs/blob/master/.travis.yml) 
+for more info.
````
info -> information

mairaw commented on docs/core/tools/using-ci-with-cli.md:
````diff
@@ -64,7 +64,17 @@ in the [CLI repo](https://github.com/dotnet/core/blob/master/Documentation/prere
+dotnet: 1.0.0-preview2-003121
+```
+
+Travis can run both `osx` (OSX 10.11) and `linux` ( Ubuntu 14.04 ) job in a build matrix, see [example .travis.yml](https://github.com/dotnet/core-docs/blob/master/.travis.yml) 
````
OSX -> OS X

am11 replied:
For accuracy, the official name of platform has been changed to macOS. :)

mairaw replied:
we've been using macOS when we refer to the platform in a generic way @am11. However, when we're talking about El Capitan, we still use OS X since that's the right branding.

mairaw commented:
Thanks @enricosada. I've removed the WIP label and left a few comments. I'll let the guys give the thumbs up on this one.

enricosada committed:
feedback

enricosada commented:
@mairaw thx for review, changed it as you suggested.

qinezh commented:
Open Publishing Build Service: The pull request content has been published and here are some sample preview links:
- Docset 1: [core-docs](https://stage.docs.microsoft.com/en-us/dotnet/core/api/system.net.sockets.protocoltype?branch=pr-en-us-460)

mairaw commented:
@blackdwarf can you finish reviewing this? Thanks!

blackdwarf commented on .travis.yml:
```diff
@@ -0,0 +1,21 @@
+
+matrix:
+  include:
+    - os: linux # Ubuntu 14.04
```
@enricosada does Travis have Xenial yet?

enricosada replied:
nope, i dont see that in documentation, just Trusty (14) and Precise (12), and no news in latest changelog https://docs.travis-ci.com/user/build-environment-updates/2016-08-24/

blackdwarf replied:
Ah, ok then. :smiley:

blackdwarf commented on .travis.yml:
```diff
@@ -0,0 +1,21 @@
+      osx_image: xcode7.2
+      dotnet: 1.0.0-preview2-003121
+
+script:
```
I'm assuming this is a sample of commands that you can run, right?

enricosada replied:
each line is a shell command, ( you can do `- which dotnet` )
travis by default run `dotnet --info` after install so the log always contains the version info
These are just a familiar smoke test, if the use know travis it's normal

blackdwarf replied:
Got it. Maybe then we should run `dotnet --version` as that will not inflate the log? I'm fine either way, whatever is best practice on Travis in general.

enricosada replied:
the `dotnet --info` it's already done by travis after the install the .net core sdk.
I'll remove it from script so it's clear is useless

mairaw commented on docs/core/tools/using-ci-with-cli.md:
```diff
@@ -64,7 +64,17 @@ in the [CLI repo](https://github.com/dotnet/core/blob/master/Documentation/prere
+for more information.
 
 ### AppVeyor
 
```
Not related to this PR: Line 81 talks about preview1. shouldn't that be preview 2? That line is also missing the period.

enricosada replied:
@mairaw good catch, fixed

enricosada committed:
fix typo

qinezh commented:
Open Publishing Build Service: The pull request content has been published and here are some sample preview links:
- Docset 1: [core-docs](https://stage.docs.microsoft.com/en-us/dotnet/core/api/system.componentmodel.dataannotations.urlattribute?branch=pr-en-us-460)

enricosada committed:
cleanup travis script

qinezh commented:
Open Publishing Build Service: The pull request content has been published and here are some sample preview links:
- Docset 1: [core-docs](https://stage.docs.microsoft.com/en-us/dotnet/core/api/system.net.http.headers.productinfoheadervalue?branch=pr-en-us-460)

mairaw commented:
LGTM. Are we good to merge @blackdwarf?

blackdwarf commented:
:shipit:

mairaw commented:
Thanks @enricosada!
